File: Python/hanoi.py
# A simple code to practice recursion solving the game of the Tower of Hanoi


def main():
    number_of_disks = int_greather_than_0("Enter a number of disks: ")
    number_of_moves = 2**number_of_disks - 1
    print("At least {} moves required to solve this game"
          .format(number_of_moves))
    hanoi(number_of_disks, 1, 2, 3)
    continue_program()


def int_greather_than_0(message):
    while True:
        try:
            _int_greather_than_0 = int(input(message))
            if _int_greather_than_0 > 0:
                break
        except ValueError:
            print("Please, enter a positive integer greater than 0")
    return _int_greather_than_0


def hanoi(number_of_disks, initial_stack, auxiliar_stack, final_stack):
    if number_of_disks == 1:
        print("Moving disk {} from stack {} to stack {}"
              .format(number_of_disks, initial_stack, final_stack))
    else:
        hanoi(number_of_disks-1, initial_stack, final_stack, auxiliar_stack)
        print("Moving disk {} from stack {} to stack {}"
              .format(number_of_disks, initial_stack, final_stack))
        hanoi(number_of_disks-1, auxiliar_stack, initial_stack, final_stack)


def continue_program():
    system_exit = input("Do you want to continue with the program? [Y/n] ")
    system_exit = system_exit.lower()
    if system_exit == "y" or system_exit == "":
        main()
    elif system_exit == "n":
        raise SystemExit
    else:
        print("You entered {}, but it is a wrong command. Good bye!"
              .format(system_exit))
        raise SystemExit


if __name__ == "__main__":
    main()

File: Java/Hanoi/src/Hanoi.java
// A simple code to practice recursion solving the game of the Tower of Hanoi

import java.util.Scanner;

public class Hanoi {

    private static boolean continueProgram = true;

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        while (continueProgram) {
            int numberOfDisks = intGreaterThan0("Enter a number of disks: ", scan);
            int numberOfMoves = (int) (Math.pow(2, numberOfDisks) - 1);
            System.out.printf("At least %d moves required to solve this game %n",
                    numberOfMoves);
            hanoi(numberOfDisks, 1, 2, 3);
            continueProgram(scan);
        }
    }

    private static int intGreaterThan0(String message, Scanner scan) {
        int intGreaterThan0 = 0;
        boolean integer = false;
        do {
            try {
                System.out.print(message);
                intGreaterThan0 = Integer.parseInt(scan.nextLine());
                integer = true;
            } catch (Exception e) {
                System.out.println("Please enter a positive integer.");
            }
            if (integer && intGreaterThan0 < 1) {
                System.out.println("Please enter a positive integer "
                        + "greater than 0.");
                integer = false;
            }
        } while (intGreaterThan0 < 1);
        return intGreaterThan0;
    }

    private static void hanoi(int numberOfDisks, int initialStack, int auxiliarStack, int finalStack) {
        if (numberOfDisks == 1) {
            System.out.printf("Moving disk %d from stack %d to stack %d%n",
                    numberOfDisks, initialStack, finalStack);
        } else {
            hanoi(numberOfDisks-1, initialStack, finalStack, auxiliarStack);
            System.out.printf("Moving disk %d from stack %d to stack %d%n",
                    numberOfDisks, initialStack, finalStack);
            hanoi(numberOfDisks-1, auxiliarStack, initialStack, finalStack);
        }
    }

    private static void continueProgram(Scanner scan) {
        System.out.print("\nDo you want to continue with the program? [Y/n] ");
        String exit = scan.nextLine();
        if ("y".equals(exit) || "".equals(exit)) {
            continueProgram = true;
        } else if ("n".equals(exit)) {
            continueProgram = false;
        } else {
            System.out.println("You entered a wrong command. Good bye!");
            continueProgram = false;
        }
    }

}
